Adds generic blacklist functionality

rff opened the pull request:
This PR adds a generic blacklist functionality that works with the
different types of checks on pmdk Makefile:

    check, pcheck, pycheck

The second commit is a list of known failing tests for ppc64le as of
today on branch devel-1.8

I decided to keep the list in a dedicated file to easier change it
without conflicts whilst we include fixes for ppc64le tests.

This feature is intended to be used on travis CI with something like
this when we add ppc64le integration in pmdk travis-ci:

    diff --git a/.travis.yml b/.travis.yml
    index d556740ad..53a64c637 100644
    --- a/.travis.yml
    +++ b/.travis.yml
    @@ -19,6 +19,7 @@ env:
         - PMDK_CXX=g++
         - REMOTE_TESTS=1
         - VALGRIND=1
    +    - BLACKLIST_FILE=ppc64.blacklist
       matrix:
         - COVERAGE=1 FAULT_INJECTION=1 TEST_BUILD=debug
         - FAULT_INJECTION=1 TEST_BUILD=debug
    diff --git a/utils/docker/build-travis.sh b/utils/docker/build-travis.sh
    index f4a1664e7..18a1644fc 100755
    --- a/utils/docker/build-travis.sh
    +++ b/utils/docker/build-travis.sh
    @@ -127,6 +127,7 @@ docker run --rm --privileged=true --name=$containerName -ti \
            --env COVERITY_SCAN_NOTIFICATION_EMAIL=$COVERITY_SCAN_NOTIFICATION_EMAIL \
            --env FAULT_INJECTION=$FAULT_INJECTION \
            --env GITHUB_REPO=$GITHUB_REPO \
    +       --env BLACKLIST_FILE=$BLACKLIST_FILE \
            $ndctl_enable \
            -v $HOST_WORKDIR:$WORKDIR \
            -v /etc/localtime:/etc/localtime \

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/4362)
<!-- Reviewable:end -->

rff commented:
I am having trouble reproducing the problem shown on travis-ci.
Whatever I use travis-ci.com or local execution (x86 and ppc64le) I cannot reproduce the error with `RUNTESTS.py`.

In a related note, I am also having trouble with any travis-ci instance, They never match exactly what pmdk uses and They fail with the source code inside docker with diferent user ownership and permissions.
Here is a sample trying to build devel-1.8 on travis-ci.com:
https://travis-ci.com/rff/pmdk/jobs/267647660

    $ ./build-travis.sh
    [sudo] password for pmdkuser: ./configure-tests.sh: line 41: /pmdk/src/test/testconfig.sh: Permission denied
    The command "./build-travis.sh" exited with 1.

It fails trying to create `testconfig.sh`

There is anything special in pmem/pmdk instance config that is not available publicity?

As said on the PR message, this effort is target to add CI support for ppc64le arch on travis.

pbalcer commented:
Everything about our CI system is public. Travis scripts are here: https://github.com/pmem/pmdk/blob/master/.travis.yml 

Sometimes the problem is that Travis will use the latest base branch for your commits, so make sure that your code is always rebased on the latest version of devel-1.8.

lucmaga commented:
Hi @pbalcer, 

> Everything about our CI system is public. Travis scripts are here: https://github.com/pmem/pmdk/blob/master/.travis.yml
> 
> Sometimes the problem is that Travis will use the latest base branch for your commits,
> so make sure that your code is always rebased on the latest version of devel-1.8.

Thanks for helping out. Me and @rff are both trying to setup travis-ci on a fork so we can test out the CI for ppc64 before creating a pull request. But we are not being able to reproduce the same results as you have here. We've tried to build both master and devel-1.8 branches without success. In both cases the problem is the "permission denied" described by @rff.

Please, let us know if you have any clue that could help out solve this.

marcinslusarz commented:
I don't know if it matters or not, but we use travis-ci.org, not travis-ci.com.

pbalcer commented:
It's hard for me to offer any advice, but I can see that travis config files are slightly different than what we use in PMDK. Maybe that has something to do with it.
Instead of setting this up manually, you could try running the tests in docker images locally, this way it's easier to debug what's going wrong.

lucmaga commented:
Thanks @pbalcer and @marcinslusarz for the tips. I found out.

Our travis settings had a Beta Feature enabled that was breaking the
build. It was enable by default when we setup the project on travis
so we didn't bother to change it.

It's working now. :)

rff commented:
@Greg09, @wlemkows, @pbalcer, @marcinslusarz, Is this still on your radar?

marcinslusarz commented on src/test/ppc64.blacklist:
```diff
@@ -0,0 +1,43 @@
+blk_nblock
```
please move this file to utils/docker to make it clear it is only for CI

rff replied:
Done. I also changed ppc64 to ppc64le, to better reflect the intended arch, and is also the arch on travis.

marcinslusarz commented on src/test/Makefile:
```diff
@@ -390,12 +401,14 @@ memcheck-summary-leaks:
 	grep "in use at exit" */memcheck*.log | grep -v " 0 bytes in 0 blocks" || true
 
 check:
+	@echo "Blacklisted tests: $(BLACKLIST_TESTS)"
```
please don't print this line when BLACKLIST_TESTS is empty

rff replied:
done

marcinslusarz commented on src/test/Makefile:
```diff
@@ -390,12 +401,14 @@ memcheck-summary-leaks:
 pycheck:
-	@./RUNTESTS.py $(subst nondebug,release,$(RUNTEST_OPTIONS))
+	@echo "Blacklisted tests: $(BLACKLIST_TESTS)"
+	@./RUNTESTS.py $(subst nondebug,release,$(RUNTEST_OPTIONS)) -- $(LOCAL_TESTS)
```
I know there are no remote Python tests, but I think this should use TESTS instead of LOCAL_TESTS.

rff replied:
done

marcinslusarz commented on src/test/Makefile:
```diff
@@ -390,12 +401,14 @@ memcheck-summary-leaks:
 
 pycheck:
-	@./RUNTESTS.py $(subst nondebug,release,$(RUNTEST_OPTIONS))
+	@echo "Blacklisted tests: $(BLACKLIST_TESTS)"
```
ditto

rff replied:
done

rff commented:
I updated the patches with:
- All @marcinslusarz requested changes,
- Renamed the blacklist from ppc64.blacklist to ppc64le.blacklist to better reflect the correct arch
- Changed commit messages to reflect the changes
- Rebased on top of devel-1.8
- Updated the ppc64le.blacklist with the failling tests as of today
